kernel/start.c

```diff
@@ -45,10 +47,36 @@ start()
   int id = r_mhartid();
   w_tp(id);
 
+  // allow access to all physical memory by S mode
+  pmpinit();
+
   // switch to supervisor mode and jump to main().
   asm volatile("mret");
 }
 
+// configures the pmp registers trivially so we can boot. it is not permitted
+// to jump to S mode without having these configured as instruction fetch will
+// fail, however we do not actually use them for protection in xv6, so we just
+// need to put something trivial there.
+//
+// see section 3.6.1 "Physical Memory Protection CSRs" in the RISC-V privileged
+// specification (v20190608)
+//
+// "If no PMP entry matches an M-mode access, the access succeeds. If no PMP
+// entry matches an S-mode or U-mode access, but at least one PMP entry is
+// implemented, the access fails." (3.6.1)
+static void
+pmpinit()
+{
+  // see figure 3.27 "PMP address register format, RV64" and table 3.10 "NAPOT
+  // range encoding in PMP address and configuration registers" in the RISC-V
+  // privileged specification
+  // we set the bits such that this matches any 56-bit physical address
+  w_pmpaddr0((~0L) >> 10);
```

Comment:
You are absolutely correct. It should be `~0ULL`. I'll change it and push a fix.

Reply:
The code is indeed accidentally functional, because the 0s in pmpaddr0[54:63] are WARL, so it is completely legit to set them to 1. I think I'll still fix it.